specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+             """Get a wrapper over the existing firewall rules.
+
+             Since this seems to be only used in Windows, it can be provided
+             only in the Windows utils.
+             The same behaviour as for :meth:`routes` can be used, that is:
+
+                 >>> rules = distro.network.firewall_rules()
+                 >>> rule = distro.network.FirewallRule(name=..., port=..., protocol=...)
+                 >>> rules.add(rule)
+                 >>> rules.delete(rule)
+                 >>> rule in rules
+                 >>> for rule in rules: print(rules)
+                 >>> del rules[i]
+                 >>> rule = rules[0]
+                 >>> rule.name, rule.port, rule.protocol, rule.allow
+
+             This gets rid of ``cloudbaseinit.osutils.firewall_add_rule`` and
+             ``cloudbaseinit.osutils.firewall_remove_rule``.
+             """
+
+         def set_static_network_config(self, adapter_name, address, netmask,
+                                       broadcast, gateway, dnsnameservers):
+             """Configure a new static network.
+
+             The :meth:``cloudinit.distros.Distro.apply_network`` should be
+             removed in the favour of this method,
+             which will be called by each network plugin.
+             The method can be a template method, providing
+             hooks for setting static DNS servers, setting static gateways or
+             setting static IP addresses, which will be implemented by specific
+             implementations of Distros.
+             """
+
+        def hosts(self):
+             """Get the content of /etc/hosts file in a more OO approach.
+
+
+             >>> hosts = distro.network.hosts()
+             >>> list(hosts) # support iteration and index access
+             >>> hosts.add(ipaddress, hostname, alias)
+             >>> hosts.delete(ipaddress, hostname, alias)
+
+             This gets rid of ``cloudinit.distros.Distro.update_etc_hosts``
+             and can provide support for adding a new hostname for Windows, as well.
+             """
+
+        class Route(object):
+             """
+             Encapsulate behaviour and state of a route.
+             Something similar to Posix can be adopted, with the following API:
+
+                  route.destination
+                  route.gateway
+                  route.flags
+                  route.refs
+                  route.use
+                  route.netif -> instance of :class:`Interface` object
+                  route.expire
+                  route.static -> 'S' in self.flags
+                  route.usable -> 'U' in self.flag
+             """
+
+          @classmethod
+          def from_route_item(self, item):
+              """
+              Build a Route from a routing entry, either from
+              the output of `netstat` or what will be used on Posix or
+              from `GetIpForwardTable`.
+              """
+
+        class RouteContainer(object):
```

Comment:
Initially, I was thinking that the container will be responsable for controlling its entries, so with RouteContainer (or whatever name we will be using) you could actually delete a route in the system itself (with route del & co). The uses seems thinner now, because we could move this behaviour in the Route itself.